src/metadata_directory/main.mo

```diff
@@ -87,37 +96,65 @@ actor class () = self {
   // API for Canister dashboard
   //
 
-  public query func find_wasm_metadata(principal : Principal, module_hash : Blob) : async WasmMetadata {
-    let ?pr = principalMap.get(storage, principal) else throw Error.reject("There is no metadata for the wasm module.");
-    let ?wasm_metadata = blobMap.get(pr.wasm_metadata_storage, module_hash) else throw Error.reject("There is no metadata for the wasm module.");
-    wasm_metadata;
+  public query func find_wasm_metadata(module_hash : Blob, principals : [Principal]) : async [(Principal, WasmMetadata)] {
+    Array.foldLeft<Principal, [(Principal, WasmMetadata)]>(
+      principals,
+      [],
+      func(acc, p) {
+        let ?pr = principalMap.get(storage, p) else return acc;
+        switch (blobMap.get(pr.wasm_metadata_storage, module_hash)) {
+          case (?wasm_metadata) Array.append(acc, [(p, wasm_metadata)]);
+          case (null) acc;
+        };
+      },
+    );
```

Comment:
We are not allowed to use Array.append in a loop because that creates quadratic complexity. Array.append copies the entire array.

Better to go through principals with a for-loop, add the entries to a Buffer, Vector or List, and at the end use toArray or Array.tabulate (once the length is known) to convert to an immutable array.

Reply:
Makes sense, I will fix it